// @nifty-oss/asset/dist/src/generated/programs/asset.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAssetProgramId = exports.getAssetProgram = exports.createAssetProgram = exports.ASSET_PROGRAM_ID = void 0;
const errors_1 = require("../errors");
exports.ASSET_PROGRAM_ID = 'AssetGtQBTSgm5s91d1RAQod5JmaZiJDxqsgtqrZud73';
function createAssetProgram() {
    return {
        name: 'asset',
        publicKey: exports.ASSET_PROGRAM_ID,
        getErrorFromCode(code, cause) {
            return (0, errors_1.getAssetErrorFromCode)(code, this, cause);
        },
        getErrorFromName(name, cause) {
            return (0, errors_1.getAssetErrorFromName)(name, this, cause);
        },
        isOnCluster() {
            return true;
        },
    };
}
exports.createAssetProgram = createAssetProgram;
function getAssetProgram(context, clusterFilter) {
    return context.programs.get('asset', clusterFilter);
}
exports.getAssetProgram = getAssetProgram;
function getAssetProgramId(context, clusterFilter) {
    return context.programs.getPublicKey('asset', exports.ASSET_PROGRAM_ID, clusterFilter);
}
exports.getAssetProgramId = getAssetProgramId;
