// @nifty-oss/asset/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAssetGpaBuilder = exports.SYSTEM_PROGRAM_ID = exports.safeFetchAsset = exports.safeFetchAllAsset = exports.fetchAsset = exports.fetchAllAsset = exports.deserializeAsset = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const generated_1 = require("./generated");
const internalAsset_1 = require("./generated/accounts/internalAsset");
const hooked_1 = require("./hooked");
__exportStar(require("./allocate"), exports);
__exportStar(require("./approve"), exports);
__exportStar(require("./burn"), exports);
__exportStar(require("./constraints"), exports);
__exportStar(require("./create"), exports);
__exportStar(require("./extensions"), exports);
__exportStar(require("./generated"), exports);
__exportStar(require("./group"), exports);
__exportStar(require("./handover"), exports);
__exportStar(require("./hooked"), exports);
__exportStar(require("./initialize"), exports);
__exportStar(require("./lock"), exports);
__exportStar(require("./mint"), exports);
__exportStar(require("./plugin"), exports);
__exportStar(require("./remove"), exports);
__exportStar(require("./resize"), exports);
__exportStar(require("./revoke"), exports);
__exportStar(require("./transfer"), exports);
__exportStar(require("./ungroup"), exports);
__exportStar(require("./unlock"), exports);
__exportStar(require("./unverify"), exports);
__exportStar(require("./update"), exports);
__exportStar(require("./updateWithBuffer"), exports);
__exportStar(require("./verify"), exports);
__exportStar(require("./write"), exports);
var internalAsset_2 = require("./generated/accounts/internalAsset");
Object.defineProperty(exports, "deserializeAsset", { enumerable: true, get: function () { return internalAsset_2.deserializeInternalAsset; } });
Object.defineProperty(exports, "fetchAllAsset", { enumerable: true, get: function () { return internalAsset_2.fetchAllInternalAsset; } });
Object.defineProperty(exports, "fetchAsset", { enumerable: true, get: function () { return internalAsset_2.fetchInternalAsset; } });
Object.defineProperty(exports, "safeFetchAllAsset", { enumerable: true, get: function () { return internalAsset_2.safeFetchAllInternalAsset; } });
Object.defineProperty(exports, "safeFetchAsset", { enumerable: true, get: function () { return internalAsset_2.safeFetchInternalAsset; } });
exports.SYSTEM_PROGRAM_ID = (0, umi_1.publicKey)('11111111111111111111111111111111');
function getAssetGpaBuilder(context) {
    const programId = context.programs.getPublicKey('asset', 'AssetGtQBTSgm5s91d1RAQod5JmaZiJDxqsgtqrZud73');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        discriminator: [0, (0, generated_1.getDiscriminatorSerializer)()],
        state: [1, (0, generated_1.getStateSerializer)()],
        standard: [2, (0, generated_1.getStandardSerializer)()],
        mutable: [3, (0, serializers_1.bool)()],
        owner: [4, (0, serializers_1.publicKey)()],
        group: [36, (0, hooked_1.getNullablePublicKeySerializer)()],
        authority: [68, (0, serializers_1.publicKey)()],
        delegate: [100, (0, hooked_1.getNullablePublicKeySerializer)()],
        roles: [132, (0, hooked_1.getDelegateRolesSerializer)()],
        name: [133, (0, serializers_1.string)({ size: 35 })],
    })
        .deserializeUsing((account) => (0, internalAsset_1.deserializeInternalAsset)(account));
}
exports.getAssetGpaBuilder = getAssetGpaBuilder;
