// @nifty-oss/asset/dist/src/update.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.update = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const extensions_1 = require("./extensions");
const generated_1 = require("./generated");
const update_1 = require("./generated/instructions/update");
function update(context, input) {
    let extension = (0, umi_1.none)();
    if (input.extension) {
        const data = (0, extensions_1.getExtensionSerializerFromType)(input.extension.type).serialize(input.extension);
        extension = {
            extensionType: input.extension.type,
            length: data.length,
            data,
        };
    }
    if (input.proxy) {
        const proxied = context.programs.clone();
        proxied.bind('asset', input.proxy);
        context = { ...context, programs: proxied };
    }
    let ix = (0, update_1.update)(context, {
        ...input,
        extension,
    });
    if (input.proxy) {
        ix = ix.addRemainingAccounts({
            pubkey: generated_1.ASSET_PROGRAM_ID,
            isWritable: false,
            isSigner: false,
        });
    }
    return ix;
}
exports.update = update;
