// @nifty-oss/asset/dist/src/generated/types/extensionType.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getExtensionTypeSerializer = exports.ExtensionType = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var ExtensionType;
(function (ExtensionType) {
    ExtensionType[ExtensionType["None"] = 0] = "None";
    ExtensionType[ExtensionType["Attributes"] = 1] = "Attributes";
    ExtensionType[ExtensionType["Blob"] = 2] = "Blob";
    ExtensionType[ExtensionType["Creators"] = 3] = "Creators";
    ExtensionType[ExtensionType["Links"] = 4] = "Links";
    ExtensionType[ExtensionType["Metadata"] = 5] = "Metadata";
    ExtensionType[ExtensionType["Grouping"] = 6] = "Grouping";
    ExtensionType[ExtensionType["Royalties"] = 7] = "Royalties";
    ExtensionType[ExtensionType["Manager"] = 8] = "Manager";
    ExtensionType[ExtensionType["Proxy"] = 9] = "Proxy";
    ExtensionType[ExtensionType["Properties"] = 10] = "Properties";
    ExtensionType[ExtensionType["Bucket"] = 11] = "Bucket";
})(ExtensionType = exports.ExtensionType || (exports.ExtensionType = {}));
function getExtensionTypeSerializer() {
    return (0, serializers_1.scalarEnum)(ExtensionType, {
        description: 'ExtensionType',
    });
}
exports.getExtensionTypeSerializer = getExtensionTypeSerializer;
