// @nifty-oss/asset/dist/src/generated/instructions/approve.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.approve = exports.getApproveInstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const shared_1 = require("../shared");
const types_1 = require("../types");
function getApproveInstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['delegateInput', (0, types_1.getDelegateInputSerializer)()],
    ], { description: 'ApproveInstructionData' }), (value) => ({ ...value, discriminator: 3 }));
}
exports.getApproveInstructionDataSerializer = getApproveInstructionDataSerializer;
// Instruction.
function approve(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('asset', 'AssetGtQBTSgm5s91d1RAQod5JmaZiJDxqsgtqrZud73');
    // Accounts.
    const resolvedAccounts = {
        asset: {
            index: 0,
            isWritable: true,
            value: input.asset ?? null,
        },
        owner: {
            index: 1,
            isWritable: false,
            value: input.owner ?? null,
        },
        delegate: {
            index: 2,
            isWritable: false,
            value: input.delegate ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.owner.value) {
        resolvedAccounts.owner.value = context.identity;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getApproveInstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.approve = approve;
