// @nifty-oss/asset/dist/src/generated/instructions/update.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.update = exports.getUpdateInstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const shared_1 = require("../shared");
const types_1 = require("../types");
function getUpdateInstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['name', (0, serializers_1.option)((0, serializers_1.string)())],
        ['mutable', (0, serializers_1.option)((0, serializers_1.bool)())],
        ['extension', (0, serializers_1.option)((0, types_1.getExtensionInputSerializer)())],
    ], { description: 'UpdateInstructionData' }), (value) => ({
        ...value,
        discriminator: 10,
        name: value.name ?? (0, umi_1.none)(),
        mutable: value.mutable ?? (0, umi_1.none)(),
        extension: value.extension ?? (0, umi_1.none)(),
    }));
}
exports.getUpdateInstructionDataSerializer = getUpdateInstructionDataSerializer;
// Instruction.
function update(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('asset', 'AssetGtQBTSgm5s91d1RAQod5JmaZiJDxqsgtqrZud73');
    // Accounts.
    const resolvedAccounts = {
        asset: {
            index: 0,
            isWritable: true,
            value: input.asset ?? null,
        },
        authority: {
            index: 1,
            isWritable: false,
            value: input.authority ?? null,
        },
        buffer: {
            index: 2,
            isWritable: true,
            value: input.buffer ?? null,
        },
        group: {
            index: 3,
            isWritable: false,
            value: input.group ?? null,
        },
        payer: {
            index: 4,
            isWritable: true,
            value: input.payer ?? null,
        },
        systemProgram: {
            index: 5,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.authority.value) {
        resolvedAccounts.authority.value = context.identity;
    }
    if (!resolvedAccounts.systemProgram.value) {
        if (resolvedAccounts.payer.value) {
            resolvedAccounts.systemProgram.value = context.programs.getPublicKey('systemProgram', '11111111111111111111111111111111');
            resolvedAccounts.systemProgram.isWritable = false;
        }
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getUpdateInstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.update = update;
